labs/intro-to-ada/chapters/standard_library.rst

```diff
@@ -0,0 +1,1073 @@
+
+:code-config:`reset_accumulator=True;accumulate_code=False`
+
+.. include:: <isopub.txt>
+
+.. role:: ada(code)
+   :language: ada
+
+.. role:: c(code)
+   :language: c
+
+.. role:: cpp(code)
+   :language: c++
+
+Simple todo list
+----------------
+
+This exercise is based on the *Simple todo list* exercise from the
+:doc:`more_about_types`. Your goal is to rewrite that exercise using
+vectors instead of arrays. Please refer to the original exercise for
+details. You may reuse the code you've implemented as a starting point.
+
+.. code:: ada lab=Standard_Library.Simple_Todo_List
+
+    --  START LAB IO BLOCK
+    in 0:Todo_List_Chk
+    out 0:TO-DO LIST Buy milk Buy tea Buy present Buy tickets Pay electricity bill Schedule dentist appointment Call sister Revise spreasheet Edit entry page Select new design Create upgrade plan
+    --  END LAB IO BLOCK
+
+    package Todo_Lists is
+
+       type Todo_Item is access String;
+
+       type Todo_List is null record;
+
+       procedure Add (Todos : in out Todo_List;
+                      Item  : String);
+
+       procedure Display (Todos : Todo_List);
+
+    end Todo_Lists;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    package body Todo_Lists is
+
+       procedure Add (Todos : in out Todo_List;
+                      Item  : String) is
+       begin
+          null;
+       end Add;
+
+       procedure Display (Todos : Todo_List) is
+       begin
+          Put_Line ("TO-DO LIST");
+       end Display;
+
+    end Todo_Lists;
+
+    with Ada.Command_Line;  use Ada.Command_Line;
+    with Ada.Text_IO;       use Ada.Text_IO;
+
+    with Todo_Lists;        use Todo_Lists;
+
+    procedure Main is
+       type Test_Case_Index is
+         (Todo_List_Chk);
+
+       procedure Check (TC : Test_Case_Index) is
+          T : Todo_List;
+       begin
+          case TC is
+             when Todo_List_Chk =>
+                Add (T, "Buy milk");
+                Add (T, "Buy tea");
+                Add (T, "Buy present");
+                Add (T, "Buy tickets");
+                Add (T, "Pay electricity bill");
+                Add (T, "Schedule dentist appointment");
+                Add (T, "Call sister");
+                Add (T, "Revise spreasheet");
+                Add (T, "Edit entry page");
+                Add (T, "Select new design");
+                Add (T, "Create upgrade plan");
+                Display (T);
+          end case;
+       end Check;
+
+    begin
+       if Argument_Count < 1 then
+          Put_Line ("ERROR: missing arguments! Exiting...");
+          return;
+       elsif Argument_Count > 1 then
+          Put_Line ("Ignoring additional arguments...");
+       end if;
+
+       Check (Test_Case_Index'Value (Argument (1)));
+    end Main;
+
+Inventory
+---------
+
+This exercise is based on the *Inventory* exercise from the
+:doc:`more_about_types`. Your goal is to rewrite that exercise using
+vectors |mdash| instead of arrays |mdash| and unbounded strings.
```

Comment:
Other than being more complicated than "to-do" list, why do we have another exercise for vectors? If we're just adding an example of "cursor" functionality, either one of these examples would work just as well with sets, and that gives the student more exposure to the different containers

Reply:
Yes, having a more complicated exercise was the main reason. However, I agree that we should have an exercise using sets, too. I'll work on that.